In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')
import datetime as dt

In [2]:
data=pd.read_csv('E:/0_1_Data Science and AI_Intellipaat/Assignments/000cap stone/Project/OnlineRetail.csv')

In [3]:
data.describe()

Quantity      UnitPrice     CustomerID
count  541909.000000  541909.000000  406829.000000
mean        9.552250       4.611114   15287.690570
std       218.081158      96.759853    1713.600303
min    -80995.000000  -11062.060000   12346.000000
25%         1.000000       1.250000   13953.000000
50%         3.000000       2.080000   15152.000000
75%        10.000000       4.130000   16791.000000
max     80995.000000   38970.000000   18287.000000

In [4]:
data.shape

(541909, 8)

In [5]:
data.nunique()

InvoiceNo      25900
StockCode       4070
Description     4223
Quantity         722
InvoiceDate    23260
UnitPrice       1630
CustomerID      4372
Country           38
dtype: int64

**Since quantity and unit price are having negative values we need to remove them since they are cancelled requests**

In [6]:
data=data[~(data['Quantity']<0)]

In [7]:
data['Amount']=data['Quantity']*data['UnitPrice']

In [8]:
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
541904    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
541905    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
541906    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
541907    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
541908    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

            InvoiceDate  UnitPrice  CustomerID         Country  Amount  
0        12/1/2010 8:26       2.55     17850.0  United Kingdom   15.30  
1        12/1/2010 8:26       3.39     17850.0  United Kingdom   20.34  
2        12/1/2010 8:26       2.75     17850.0  United Kingdom   22.00  
3        12/1/2010 8:26       3.39     17850.0  United Kingdom   20.34  
4        12/1/2010 8:26       3.39     17850.0  United Kingdom   20.34  
...                 ...        ...         ...             ...     ...  
541904  12/9/2011 12:50       0.85     12680.0          France   10.20  
541905  12/9/2011 12:50       2.10     12680.0          France   12.60  
541906  12/9/2011 12:50       4.15     12680.0          France   16.60  
541907  12/9/2011 12:50       4.15     12680.0          France   16.60  
541908  12/9/2011 12:50       4.95     12680.0          France   14.85  

[531285 rows x 9 columns]

In [9]:
data=data[data['Amount']>0]

In [10]:
data.describe(include='all')

InvoiceNo StockCode                         Description       Quantity  \
count     530104    530104                              530104  530104.000000   
unique     19960      3922                                4026            NaN   
top       573585    85123A  WHITE HANGING HEART T-LIGHT HOLDER            NaN   
freq        1114      2265                                2323            NaN   
mean         NaN       NaN                                 NaN      10.542037   
std          NaN       NaN                                 NaN     155.524124   
min          NaN       NaN                                 NaN       1.000000   
25%          NaN       NaN                                 NaN       1.000000   
50%          NaN       NaN                                 NaN       3.000000   
75%          NaN       NaN                                 NaN      10.000000   
max          NaN       NaN                                 NaN   80995.000000   

             InvoiceDate      UnitPrice     CustomerID         Country  \
count             530104  530104.000000  397884.000000          530104   
unique             18499            NaN            NaN              38   
top     10/31/2011 14:41            NaN            NaN  United Kingdom   
freq                1114            NaN            NaN          485123   
mean                 NaN       3.907625   15294.423453             NaN   
std                  NaN      35.915681    1713.141560             NaN   
min                  NaN       0.001000   12346.000000             NaN   
25%                  NaN       1.250000   13969.000000             NaN   
50%                  NaN       2.080000   15159.000000             NaN   
75%                  NaN       4.130000   16795.000000             NaN   
max                  NaN   13541.330000   18287.000000             NaN   

               Amount  
count   530104.000000  
unique            NaN  
top               NaN  
freq              NaN  
mean        20.121871  
std        270.356743  
min          0.001000  
25%          3.750000  
50%          9.900000  
75%         17.700000  
max     168469.600000

**Converting Invoice date to datetime object**

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 530104 entries, 0 to 541908
Data columns (total 9 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    530104 non-null  object 
 1   StockCode    530104 non-null  object 
 2   Description  530104 non-null  object 
 3   Quantity     530104 non-null  int64  
 4   InvoiceDate  530104 non-null  object 
 5   UnitPrice    530104 non-null  float64
 6   CustomerID   397884 non-null  float64
 7   Country      530104 non-null  object 
 8   Amount       530104 non-null  float64
dtypes: float64(3), int64(1), object(5)
memory usage: 40.4+ MB


In [12]:
data['InvoiceDate']=pd.to_datetime(data['InvoiceDate'])
data['CustomerID'] = data['CustomerID'].astype(float)

In [13]:
data.reset_index(drop=True,inplace=True)
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
530099    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
530100    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
530101    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
530102    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
530103    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  Amount  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom   15.30  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom   22.00  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34  
...                    ...        ...         ...             ...     ...  
530099 2011-12-09 12:50:00       0.85     12680.0          France   10.20  
530100 2011-12-09 12:50:00       2.10     12680.0          France   12.60  
530101 2011-12-09 12:50:00       4.15     12680.0          France   16.60  
530102 2011-12-09 12:50:00       4.15     12680.0          France   16.60  
530103 2011-12-09 12:50:00       4.95     12680.0          France   14.85  

[530104 rows x 9 columns]

In [14]:
data.isnull().sum()

InvoiceNo           0
StockCode           0
Description         0
Quantity            0
InvoiceDate         0
UnitPrice           0
CustomerID     132220
Country             0
Amount              0
dtype: int64

**Since the missing values are in customer id column alone its better to drop them as the problem in hand is about customer segmentation**

In [15]:
data.dropna(inplace=True)
data.reset_index(inplace=True,drop=True)
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
397879    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
397880    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
397881    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
397882    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
397883    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  Amount  
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom   15.30  
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34  
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom   22.00  
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34  
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34  
...                    ...        ...         ...             ...     ...  
397879 2011-12-09 12:50:00       0.85     12680.0          France   10.20  
397880 2011-12-09 12:50:00       2.10     12680.0          France   12.60  
397881 2011-12-09 12:50:00       4.15     12680.0          France   16.60  
397882 2011-12-09 12:50:00       4.15     12680.0          France   16.60  
397883 2011-12-09 12:50:00       4.95     12680.0          France   14.85  

[397884 rows x 9 columns]

In [16]:
data.isnull().sum()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
Amount         0
dtype: int64

# Customer segmentation based on:
1. Recency : Number of days since last purchase
2. Frequency : Number of tracsactions
3. Monetary : Total amount of transactions (revenue contributed)

**Monetary**

In [17]:
rfm_m=data.groupby('CustomerID')['Amount'].sum()
rfm_m=rfm_m.reset_index()
rfm_m

CustomerID    Amount
0        12346.0  77183.60
1        12347.0   4310.00
2        12348.0   1797.24
3        12349.0   1757.55
4        12350.0    334.40
...          ...       ...
4333     18280.0    180.60
4334     18281.0     80.82
4335     18282.0    178.05
4336     18283.0   2094.88
4337     18287.0   1837.28

[4338 rows x 2 columns]

**Frequency**

In [18]:
rfm_f=data.groupby('CustomerID')['InvoiceNo'].count()
rfm_f=rfm_f.reset_index()
rfm_f

CustomerID  InvoiceNo
0        12346.0          1
1        12347.0        182
2        12348.0         31
3        12349.0         73
4        12350.0         17
...          ...        ...
4333     18280.0         10
4334     18281.0          7
4335     18282.0         12
4336     18283.0        756
4337     18287.0         70

[4338 rows x 2 columns]

**Recency**

In [19]:
rec_date=max(data['InvoiceDate'])
data['Difference']=rec_date-data['InvoiceDate']
data

InvoiceNo StockCode                          Description  Quantity  \
0         536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1         536365     71053                  WHITE METAL LANTERN         6   
2         536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3         536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4         536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
...          ...       ...                                  ...       ...   
397879    581587     22613          PACK OF 20 SPACEBOY NAPKINS        12   
397880    581587     22899         CHILDREN'S APRON DOLLY GIRL          6   
397881    581587     23254        CHILDRENS CUTLERY DOLLY GIRL          4   
397882    581587     23255      CHILDRENS CUTLERY CIRCUS PARADE         4   
397883    581587     22138        BAKING SET 9 PIECE RETROSPOT          3   

               InvoiceDate  UnitPrice  CustomerID         Country  Amount  \
0      2010-12-01 08:26:00       2.55     17850.0  United Kingdom   15.30   
1      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34   
2      2010-12-01 08:26:00       2.75     17850.0  United Kingdom   22.00   
3      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34   
4      2010-12-01 08:26:00       3.39     17850.0  United Kingdom   20.34   
...                    ...        ...         ...             ...     ...   
397879 2011-12-09 12:50:00       0.85     12680.0          France   10.20   
397880 2011-12-09 12:50:00       2.10     12680.0          France   12.60   
397881 2011-12-09 12:50:00       4.15     12680.0          France   16.60   
397882 2011-12-09 12:50:00       4.15     12680.0          France   16.60   
397883 2011-12-09 12:50:00       4.95     12680.0          France   14.85   

              Difference  
0      373 days 04:24:00  
1      373 days 04:24:00  
2      373 days 04:24:00  
3      373 days 04:24:00  
4      373 days 04:24:00  
...                  ...  
397879   0 days 00:00:00  
397880   0 days 00:00:00  
397881   0 days 00:00:00  
397882   0 days 00:00:00  
397883   0 days 00:00:00  

[397884 rows x 10 columns]

In [20]:
rfm_r=data.groupby('CustomerID')['Difference'].min()
rfm_r=rfm_r.reset_index()
rfm_r

CustomerID        Difference
0        12346.0 325 days 02:49:00
1        12347.0   1 days 20:58:00
2        12348.0  74 days 23:37:00
3        12349.0  18 days 02:59:00
4        12350.0 309 days 20:49:00
...          ...               ...
4333     18280.0 277 days 02:58:00
4334     18281.0 180 days 01:57:00
4335     18282.0   7 days 01:07:00
4336     18283.0   3 days 00:48:00
4337     18287.0  42 days 03:21:00

[4338 rows x 2 columns]

In [21]:
rfm_r['Difference']=rfm_r['Difference'].dt.days

In [22]:
rfm_r

CustomerID  Difference
0        12346.0         325
1        12347.0           1
2        12348.0          74
3        12349.0          18
4        12350.0         309
...          ...         ...
4333     18280.0         277
4334     18281.0         180
4335     18282.0           7
4336     18283.0           3
4337     18287.0          42

[4338 rows x 2 columns]

**Combining all the three features into one single dataframe**

In [23]:
rfm=rfm_m.merge(rfm_f,how='inner',on='CustomerID')
rfm=rfm.merge(rfm_r,how='inner',on='CustomerID')
rfm.columns=['CustomerID','Monetary','Frequency','Recency']
rfm

CustomerID  Monetary  Frequency  Recency
0        12346.0  77183.60          1      325
1        12347.0   4310.00        182        1
2        12348.0   1797.24         31       74
3        12349.0   1757.55         73       18
4        12350.0    334.40         17      309
...          ...       ...        ...      ...
4333     18280.0    180.60         10      277
4334     18281.0     80.82          7      180
4335     18282.0    178.05         12        7
4336     18283.0   2094.88        756        3
4337     18287.0   1837.28         70       42

[4338 rows x 4 columns]

# Customer segmentation based on RFM

In [24]:
rfm.describe()

CustomerID       Monetary    Frequency      Recency
count   4338.000000    4338.000000  4338.000000  4338.000000
mean   15300.408022    2054.266460    91.720609    91.536422
std     1721.808492    8989.230441   228.785094   100.014169
min    12346.000000       3.750000     1.000000     0.000000
25%    13813.250000     307.415000    17.000000    17.000000
50%    15299.500000     674.485000    41.000000    50.000000
75%    16778.750000    1661.740000   100.000000   141.000000
max    18287.000000  280206.020000  7847.000000   373.000000

In [25]:
rfm.nunique()

CustomerID    4338
Monetary      4284
Frequency      460
Recency        349
dtype: int64

In [26]:
# 5 is best score and 1 being the worst
rfm['Monetary_score']=pd.qcut(rfm['Monetary'], 5, labels = [1, 2, 3, 4, 5])
rfm['Frequency_score']=pd.qcut(rfm['Frequency'].rank(method="first"),5,labels=[1,2,3,4,5])
rfm['Recency_score']=pd.qcut(rfm['Recency'].rank(method="first"),5,labels=[5,4,3,2,1])
rfm['rfm_score']=rfm['Recency_score'].astype(str)+rfm['Frequency_score'].astype(str)+rfm['Monetary_score'].astype(str)
rfm

CustomerID  Monetary  Frequency  Recency Monetary_score Frequency_score  \
0        12346.0  77183.60          1      325              5               1   
1        12347.0   4310.00        182        1              5               5   
2        12348.0   1797.24         31       74              4               3   
3        12349.0   1757.55         73       18              4               4   
4        12350.0    334.40         17      309              2               2   
...          ...       ...        ...      ...            ...             ...   
4333     18280.0    180.60         10      277              1               1   
4334     18281.0     80.82          7      180              1               1   
4335     18282.0    178.05         12        7              1               1   
4336     18283.0   2094.88        756        3              5               5   
4337     18287.0   1837.28         70       42              4               4   

     Recency_score rfm_score  
0                1       115  
1                5       555  
2                2       234  
3                4       444  
4                1       122  
...            ...       ...  
4333             1       111  
4334             1       111  
4335             5       511  
4336             5       555  
4337             3       344  

[4338 rows x 8 columns]

In [31]:
rfm['rfm_score']=rfm['rfm_score'].astype(int)
rfm['Status']=np.zeros((4338,1))

In [44]:
for i in range(0,4338):
    if rfm['rfm_score'][i] in [111, 112, 121, 131,141,151,142]:
        rfm['Status'][i]='Lost customers'
        
    if rfm['rfm_score'][i] in [332, 322, 233, 232, 223, 222, 132, 123, 122, 212, 211]:
        rfm['Status'][i]='Hibernating customers'
        
    if rfm['rfm_score'][i] in [155, 154, 144, 214,215,115, 114, 113]:
        rfm['Status'][i]='Cannot Lose Them'
        
    if rfm['rfm_score'][i] in [255, 254, 245, 244, 253, 252, 243, 242, 235, 234, 225, 224, 153, 152, 145, 143, 135, 134, 133, 125, 124]:
        rfm['Status'][i]='At Risk'
        
    if rfm['rfm_score'][i] in [331, 321, 312, 221, 213, 231, 241, 251]:
        rfm['Status'][i]='About To Sleep'
        
    if rfm['rfm_score'][i] in [535, 534, 443, 434, 343, 334, 325, 324]:
        rfm['Status'][i]='Need Attention'
        
    if rfm['rfm_score'][i] in [525, 524, 523, 522, 521, 515, 514, 513, 425,424, 413,414,415, 315, 314, 313]:
        rfm['Status'][i]='Promising'
        
    if rfm['rfm_score'][i] in [512, 511, 422, 421, 412, 411, 311]:
        rfm['Status'][i]='New Customers'
        
    if rfm['rfm_score'][i] in [553, 551, 552, 541, 542, 533, 532, 531, 452, 451, 442, 441, 431, 453, 433, 432, 423, 353, 352, 351, 342, 341, 333, 323]:
        rfm['Status'][i]='Potential Loyalist'
        
    if rfm['rfm_score'][i] in [543, 444, 435, 355, 354, 345, 344, 335]:
        rfm['Status'][i]='Loyal'
        
    if rfm['rfm_score'][i] in [555, 554, 544, 545, 454, 455, 445]:
        rfm['Status'][i]='Champions'

In [45]:
rfm

CustomerID  Monetary  Frequency  Recency Monetary_score Frequency_score  \
0        12346.0  77183.60          1      325              5               1   
1        12347.0   4310.00        182        1              5               5   
2        12348.0   1797.24         31       74              4               3   
3        12349.0   1757.55         73       18              4               4   
4        12350.0    334.40         17      309              2               2   
...          ...       ...        ...      ...            ...             ...   
4333     18280.0    180.60         10      277              1               1   
4334     18281.0     80.82          7      180              1               1   
4335     18282.0    178.05         12        7              1               1   
4336     18283.0   2094.88        756        3              5               5   
4337     18287.0   1837.28         70       42              4               4   

     Recency_score  rfm_score                 Status  
0                1        115       Cannot Lose Them  
1                5        555              Champions  
2                2        234                At Risk  
3                4        444                  Loyal  
4                1        122  Hibernating customers  
...            ...        ...                    ...  
4333             1        111         Lost customers  
4334             1        111         Lost customers  
4335             5        511          New Customers  
4336             5        555              Champions  
4337             3        344                  Loyal  

[4338 rows x 9 columns]

**Summary of the status**

In [46]:
status=pd.DataFrame(rfm.groupby('Status')['CustomerID'].count())
status.reset_index(inplace=True)
status['Percentage']=np.zeros((11,1))
for i in range (0,11):
    status['Percentage'][i]=(status['CustomerID'][i]/np.sum(status['CustomerID']))*100
status.columns=['Status','Number of Customers','Percentage of customers']

In [47]:
status

Status  Number of Customers  Percentage of customers
0          About To Sleep                  192                 4.426003
1                 At Risk                  411                 9.474412
2        Cannot Lose Them                   89                 2.051637
3               Champions                  795                18.326418
4   Hibernating customers                  810                18.672199
5          Lost customers                  450                10.373444
6                   Loyal                  420                 9.681881
7          Need Attention                  226                 5.209774
8           New Customers                  302                 6.961734
9      Potential Loyalist                  510                11.756570
10              Promising                  133                 3.065929

**At risk** customers

In [48]:
rfm[rfm['Status']=='At Risk']

#At Risk Customers are customers who purchased often and spent big amounts, but haven’t purchased recently.

CustomerID  Monetary  Frequency  Recency Monetary_score Frequency_score  \
2        12348.0   1797.24         31       74              4               3   
7        12354.0   1079.40         58      231              4               3   
27       12378.0   4008.62        219      129              5               5   
42       12399.0   1108.65         59      118              4               4   
46       12405.0   1710.39         54      148              4               3   
...          ...       ...        ...      ...            ...             ...   
4271     18196.0    689.13        141       94              3               5   
4302     18235.0   1796.48        104       71              4               4   
4308     18242.0   2232.49         78       71              5               4   
4314     18251.0   4314.72         16       86              5               2   
4319     18260.0   2643.20        134      172              5               5   

     Recency_score  rfm_score   Status  
2                2        234  At Risk  
7                1        134  At Risk  
27               2        255  At Risk  
42               2        244  At Risk  
46               2        234  At Risk  
...            ...        ...      ...  
4271             2        253  At Risk  
4302             2        244  At Risk  
4308             2        245  At Risk  
4314             2        225  At Risk  
4319             2        255  At Risk  

[411 rows x 9 columns]

In [58]:
print(rfm[rfm['Status']=='At Risk']['Recency'].max(),rfm[rfm['Status']=='At Risk']['Recency'].min())
rfm[rfm['Status']=='At Risk'].describe()

371 71


CustomerID      Monetary   Frequency     Recency   rfm_score
count    411.000000    411.000000  411.000000  411.000000  411.000000
mean   15228.559611   1596.165669   81.922141  147.827251  213.841849
std     1740.499624   1401.879791   60.912562   73.595280   48.201982
min    12348.000000    278.740000   14.000000   71.000000  124.000000
25%    13687.000000    781.190000   44.000000   88.000000  145.000000
50%    15303.000000   1145.840000   66.000000  123.000000  242.000000
75%    16730.000000   1832.565000   98.500000  189.000000  244.000000
max    18260.000000  11072.670000  543.000000  371.000000  255.000000

The customers at risk have not visited the shop for atleast 71 days and atmost 371 days.

There are totally 411 customers with an average spending of 1596.16 dollars.

On an average they have made 82 purchases.

We can send them personalized reactivation campaigns to reconnect or offer renewals and helpful products to encourage another purchase.

**Cannot lose** customers

In [59]:
rfm[rfm['Status']=='Cannot Lose Them']

#They are customers who used to visit and purchase quite often, but haven’t been visiting recently. 
#If we don't take any actions immediately, we will lose them.

CustomerID  Monetary  Frequency  Recency Monetary_score Frequency_score  \
0        12346.0  77183.60          1      325              5               1   
26       12377.0   1628.12         77      314              4               4   
31       12383.0   1850.56         99      184              4               4   
120      12497.0   1043.10         14       81              4               1   
123      12501.0   2169.39        149      336              5               5   
...          ...       ...        ...      ...            ...             ...   
4224     18133.0    931.50          1      211              3               1   
4270     18194.0    973.84          7       81              4               1   
4274     18202.0   1396.99         68      199              4               4   
4299     18231.0   2083.27        124      191              5               5   
4305     18239.0   1561.61         88      218              4               4   

     Recency_score  rfm_score            Status  
0                1        115  Cannot Lose Them  
26               1        144  Cannot Lose Them  
31               1        144  Cannot Lose Them  
120              2        214  Cannot Lose Them  
123              1        155  Cannot Lose Them  
...            ...        ...               ...  
4224             1        113  Cannot Lose Them  
4270             2        214  Cannot Lose Them  
4274             1        144  Cannot Lose Them  
4299             1        155  Cannot Lose Them  
4305             1        144  Cannot Lose Them  

[89 rows x 9 columns]

In [60]:
rfm[rfm['Status']=='Cannot Lose Them'].describe()

CustomerID      Monetary   Frequency     Recency   rfm_score
count     89.000000     89.000000   89.000000   89.000000   89.000000
mean   15162.471910   3335.147652   51.932584  217.595506  147.101124
std     1769.721082  10035.792985   63.485288   75.476625   36.472606
min    12346.000000    516.420000    1.000000   73.000000  113.000000
25%    13715.000000    973.840000    5.000000  184.000000  114.000000
50%    15069.000000   1289.500000   12.000000  217.000000  144.000000
75%    16714.000000   1850.560000   91.000000  268.000000  155.000000
max    18239.000000  77183.600000  297.000000  371.000000  215.000000

There are totally 89 customers who have not visited for atleast 73 days and atmost 371 days

They have spent an average of 3335 dollars

On an average they have made 52 purchases

We can bring them back with relevant promotions, seasonal discounts to make them feel special or 
run surveys to find out what went wrong and avoid losing them to a competitor.


**New** customers

In [62]:
rfm[rfm['Status']=='New Customers']

#They are customers who have a high overall RFM score but are not frequent shoppers.

CustomerID  Monetary  Frequency  Recency Monetary_score Frequency_score  \
19       12367.0    168.90         11        3              1               1   
78       12442.0    172.06         12        2              1               1   
80       12445.0    133.40          4       21              1               1   
87       12452.0    430.57         22       16              2               2   
121      12498.0    403.70         23       20              2               2   
...          ...       ...        ...      ...            ...             ...   
4312     18249.0     95.34          8       16              1               1   
4328     18273.0    204.00          3        1              1               1   
4329     18274.0    175.92         11       29              1               1   
4331     18277.0    110.38          8       57              1               1   
4335     18282.0    178.05         12        7              1               1   

     Recency_score  rfm_score         Status  
19               5        511  New Customers  
78               5        511  New Customers  
80               4        411  New Customers  
87               4        422  New Customers  
121              4        422  New Customers  
...            ...        ...            ...  
4312             4        411  New Customers  
4328             5        511  New Customers  
4329             4        411  New Customers  
4331             3        311  New Customers  
4335             5        511  New Customers  

[302 rows x 9 columns]

In [63]:
rfm[rfm['Status']=='New Customers'].describe()

CustomerID    Monetary   Frequency     Recency   rfm_score
count    302.000000  302.000000  302.000000  302.000000  302.000000
mean   15311.711921  210.105861   11.102649   27.841060  400.248344
std     1724.493572  111.893360    6.907509   17.675778   66.162047
min    12367.000000    6.200000    1.000000    0.000000  311.000000
25%    13819.750000  125.362500    6.000000   16.000000  311.000000
50%    15277.500000  187.030000   10.000000   23.000000  411.500000
75%    16846.750000  284.632500   15.000000   38.000000  422.000000
max    18282.000000  486.720000   29.000000   71.000000  512.000000

There are totally 302 customers who have visited atleast 1 day ago and atmost 71 days before

On an average they have made 11 purchases

They have spent 210 dollars on an averagee

We can start building relationships with these customers by providing onboarding support and special offers to increase their visits.

**Need attention** customers

In [64]:
rfm[rfm['Status']=='Need Attention']

# They don't frequently purchase, but they spend quite a good amount of money

CustomerID  Monetary  Frequency  Recency Monetary_score Frequency_score  \
22       12372.0   1298.04         52       71              4               3   
38       12394.0   1272.48         27       63              4               2   
81       12446.0   1002.31         57       57              4               3   
89       12454.0   3528.34         15       55              5               2   
93       12458.0    947.40         38       70              4               3   
...          ...       ...        ...      ...            ...             ...   
4231     18144.0   2888.75         43        7              5               3   
4247     18167.0   1384.86         37       22              4               3   
4250     18170.0    980.31         38       32              4               3   
4251     18171.0   1231.31         54       58              4               3   
4315     18252.0    526.67         98       50              3               4   

     Recency_score  rfm_score          Status  
22               3        334  Need Attention  
38               3        324  Need Attention  
81               3        334  Need Attention  
89               3        325  Need Attention  
93               3        334  Need Attention  
...            ...        ...             ...  
4231             5        535  Need Attention  
4247             4        434  Need Attention  
4250             3        334  Need Attention  
4251             3        334  Need Attention  
4315             3        343  Need Attention  

[226 rows x 9 columns]

In [66]:
rfm[rfm['Status']=='Need Attention'].describe()

CustomerID      Monetary   Frequency     Recency   rfm_score
count    226.000000    226.000000  226.000000  226.000000  226.000000
mean   15250.123894   1470.642389   57.252212   31.610619  414.238938
std     1719.928190   2533.784307   24.893428   20.336692   78.559223
min    12372.000000    490.520000   15.000000    0.000000  324.000000
25%    13730.250000    808.302500   40.000000   15.250000  334.000000
50%    15180.500000   1054.615000   51.000000   29.000000  434.000000
75%    16790.500000   1384.162500   72.000000   49.750000  443.000000
max    18252.000000  26879.040000  121.000000   71.000000  535.000000

There are 226 customers who have purchased atleast 15 days ago and atmost 121 days ago

They have spent 1470 dollars on an average and haev purchased 57 times on average

In order to make them transform into a customer that purchases frequently, we can offer them some discount with a time limit of 30 days. So that they would revisit and purchase.

In [67]:
rfm.to_csv('rfm')